### PR TITLE
Refactor `StringValidator` to avoid inheritance

### PR DESCRIPTION
**Pull request by vbreuss**
Refactor the `StringValidator` to favor composition over inheritance (see [this comment](https://github.com/fluentassertions/fluentassertions/issues/2050#issuecomment-1416710889)):
Instead of subclassing the abstract `StringValidator` class, this class now accepts an interface `IStringMismatchValidator` which provides the necessary implementation details.
If this PR is merged, it should simplify implementing #2050. 

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: Src/FluentAssertions/Primitives/IStringMismatchValidator.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using FluentAssertions.Execution;
+
+namespace FluentAssertions.Primitives;
+
+internal interface IStringMismatchValidator
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/IStringMismatchValidator.cs:5]: 🔧 This interface needs some documentation
🤔 What about `IStringComparisonStrategy` to emphase we're using a Strategy Pattern here?

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Primitives/IStringMismatchValidator.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/IStringMismatchValidator.cs]: > This interface needs some documentation

I meant both the interface _and_ its properties ;-)

Comment (vbreuss) [reply Src/FluentAssertions/Primitives/IStringMismatchValidator.cs]: Sorry, it seems I was too quick :-)

File: Src/FluentAssertions/Primitives/NegatedStringStartStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    public void ValidateAgainstMismatch(IAssertionScope assertion, string subject, string expected)
+    {
+        bool isMatch = subject.StartsWith(expected, stringComparison);
+
+        if (isMatch)
+        {
+            assertion.FailWith(ExpectationDescription + "{0}{reason}, but found {1}.",
+                expected, subject);
+        }
```
Comment (IT-VBFK) [inline Src/FluentAssertions/Primitives/NegatedStringStartStrategy.cs:35]: Why not just use the internal FA-API?

like

```csharp
assertion
    .ForCondition(subject.StartsWith(expected, stringComparison))
    .FailWith(ExpectationDescription + "{0}{reason}, but found {1}.",
                 expected, subject);
```

Comment (vbreuss) [reply Src/FluentAssertions/Primitives/NegatedStringStartStrategy.cs]: No reason, I just didn't refactor the existing code :-)
image

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/NegatedStringStartStrategy.cs]: And that code pre-dates the `ForCondition` construct ;-)

File: Src/FluentAssertions/Primitives/StringWildcardMatchingStrategy.cs [hunk 3]
```diff
@@ -2,39 +2,39 @@
...
 
-        if (isMatch && Negate)
+        if (Negate)
         {
-            Assertion.FailWith(ExpectationDescription + "but {1} matches.", Expected, Subject);
+            assertion.FailWith(ExpectationDescription + "but {1} matches.", expected, subject);
+        }
+        else
+        {
+            assertion.FailWith(ExpectationDescription + "but {1} does not.", expected, subject);
```
Comment (IT-VBFK) [inline Src/FluentAssertions/Primitives/StringWildcardMatchingStrategy.cs:26]: Internal FA-API:
Applies to other spots as well ;)

Comment (vbreuss) [reply Src/FluentAssertions/Primitives/StringWildcardMatchingStrategy.cs]: You are right. Originally I tried to change as little as possible, but will refactor more thoroughly now :-)

**Comment by IT-VBFK** [comment]
Something to rethink either: I think it is odd to have a `StringStartStrategy` and a seperate `NegatedStringStartStrategy` but do not have this pattern in the other `...Strategy`s.

What about aligning them either use into `Xxx` and `NegatedXxx` files or handle it like the `StringWildcardStrategy`?

**Comment by vbreuss** [comment]
> Something to rethink either: I think it is odd to have a `StringStartStrategy` and a seperate `NegatedStringStartStrategy` but do not have this pattern in the other `...Strategy`s.
> 
> What about aligning them either use into `Xxx` and `NegatedXxx` files or handle it like the `StringWildcardStrategy`?
> 
> If we agree to the latter: Should then `Negated` and `IgnoreCase` go into the contract `IStringComparisonStrategy`?

The `StringStartStrategy` and `NegatedStringStartStrategy` are implemented very differently and differ in the explicit failure messages for different scenarios.
I don't think it a good idea to also change the behavior as part of a "refactor"-PR and would rather adapt this in a separate pull request.

**Comment by dennisdoomen** [comment]
I believe that enough work is done and there's no need to increase the scope.  Moreover, the involved classes are all `internal` and thus easy to change later on

**Review by dennisdoomen — APPROVED** [review]

File: Src/FluentAssertions/Primitives/StringWildcardMatchingStrategy.cs [hunk 4]
```diff
@@ -2,39 +2,32 @@
...
-        if (isMatch && Negate)
-        {
-            Assertion.FailWith(ExpectationDescription + "but {1} matches.", Expected, Subject);
-        }
+        assertion
+            .ForCondition(!Negate || !isMatch)
+            .FailWith(ExpectationDescription + "but {1} matches.", expected, subject)
+            .Then
+            .ForCondition(Negate || isMatch)
+            .FailWith(ExpectationDescription + "but {1} does not.", expected, subject);
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/StringWildcardMatchingStrategy.cs:20]: Let's keep this method as it was.
I haven't dug into the history, but I suspect it was constructed as it was for performance reasons.

File: Src/FluentAssertions/Primitives/StringStartStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        assertion
+            .ForCondition(subject.Length >= expected.Length)
+            .FailWith(ExpectationDescription + "{0}{reason}, but {1} is too short.", expected, subject)
+            .Then
+            .ForCondition(subject.StartsWith(expected, stringComparison))
+            .FailWith(ExpectationDescription + "{0}{reason}, but {1} differs near " +
+                subject.IndexedSegmentAt(subject.IndexOfFirstMismatch(expected, stringComparison)) +
+                ".",
+                expected, subject);
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/StringStartStrategy.cs:38]: This will now also invoke `IndexedSegmentAt` and `IndexOfFirstMismatch` on success.

**Review by jnyrup — COMMENTED** [review]
That's some nice untangling of this unnecessary class hierarchy 🧹

File: Src/FluentAssertions/Primitives/StringStartStrategy.cs (hunk 5 above)
Comment (vbreuss) [reply Src/FluentAssertions/Primitives/StringStartStrategy.cs]: Changed in [b34b47b](https://github.com/fluentassertions/fluentassertions/pull/2296/commits/b34b47b567e800470868e19b80901b4ee013f0c6).

File: Src/FluentAssertions/Primitives/StringWildcardMatchingStrategy.cs (hunk 4 above)
Comment (vbreuss) [reply Src/FluentAssertions/Primitives/StringWildcardMatchingStrategy.cs]: Changed in [b34b47b](https://github.com/fluentassertions/fluentassertions/pull/2296/commits/b34b47b567e800470868e19b80901b4ee013f0c6)

**Review by jnyrup — APPROVED** [review]

**Pull request merged**